### DISCUSSION TITLE
Another idea for the interface of the diff algorithm

### DISCUSSION THREAD

**Discussion by richter-mvtec** [arximboldi/immer#203] [Ideas]
Hi,

I had another idea about the interface of the diff algorithm.
I wonder, if it's better or more C++-like to provide a range instead of callbacks:
```
struct change<KeyType, ValueType> {
  KeyType const& key, // key that changed
  ValueType const* a, // pointer to data in a if present
  ValueType const* b  // pointer to data in b if present
}
for (auto const& change : immer::diff(map_a, map_b))
{
  if (a != nullptr && b != nullptr)
    // change
  else if (b == nullptr)
    // new value b
  else
    // removed in b
}
```
We could also add an enum to the struct to distinguish added/changed/removed.
I had the idea, because I wanted only the keys of the items that changed in a immer::map.
What do you think?

Best wishes,
Florian

**Comment by arximboldi** [comment]
While I agree that a range-like interface is more C++-ish and more composable with the Ranges in C++20, I find the current interface more efficient. This is why Immer provides its own `for_each` that is more efficient than using a `for`-loop. You can also see the complexity of the implementation of iterators for Immer data-strucutres to guess that maybe it would not be that easy to implement the suggested interface... 

Except in C++20, maybe they could be implemented more easily using coroutines. In that case, maybe it could be a C++20 add-on, or even something that people just do in their code-base. Sadly I am not on C++20 yet on my main projects now to experiment.
